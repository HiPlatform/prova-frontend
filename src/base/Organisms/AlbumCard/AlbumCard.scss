@import "../../../variables";

.albumCard {
  display: flex;
  width: 142px;
  height: 340px;
  flex-direction: column;
  margin: 10px 4px;
  align-items: stretch;
  background-color: $cleanBackground;
  border-bottom: 1px solid $borderColor;
  border-radius: 5px;
  box-shadow: $shadow;
  flex-grow: 1;
  max-width: 284px;

  & > img {
    width: 100%;
    height: 160px;
  }

  &__info {
    width: 100%;
    height: 100%;
    position: relative;
    padding: 6px;

    &__title > .attributeField, &__extra > .attributeField {
      margin: 4px 0;
      padding-right: 4px;
      flex: 1 auto;
    }

    &__title {
      font-size: 13px;
      font-weight: bold;
    }

    &__extra {
      font-size: 11px;
    }

    &__actions {
      width: 100%;
      position: absolute;
      bottom: 6px;
    }
  }
}