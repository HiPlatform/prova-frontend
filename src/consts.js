export const URLS = {
  SEARCH: 'https://api.spotify.com/v1/search',
};

export const CREDENTIALS = {
  headers: {
    Authorization: 'Bearer BQDGGAag6Hocue0iXDFWp0O9CdmjCSq6nI9R2-HE-_N-skmHG8V3btyKOZnilxWjn7cpkFMAbM70G8CH_-Y',
  },
};

export const HOMEPAGE_ACTIONS = {
  SEARCH_DATA: 'SEARCH_DATA',
  SEARCH_TYPE: 'SEARCH_TYPE',
  SEARCH_CHANGE: 'SEARCH_CHANGE',
  SEARCH_DETAILS: 'SEARCH_DETAILS',
  NAVIGATE: 'NAVIGATE',
};
