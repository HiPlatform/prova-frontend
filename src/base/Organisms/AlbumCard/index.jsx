/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import { pathOr } from 'ramda';

import Button from '../../Atoms/Button';
import Image from '../../Atoms/Image';
import AttributeField from '../../Molecules/AttributeField';

import './AlbumCard.css';

const checkArtists = items =>
  (items.length > 1
    ? 'Various artists'
    : pathOr('', ['0', 'name'], items));

const checkAvailable = (country, markets) =>
  (markets.indexOf(country) === -1
    ? 'Unavailable in your country'
    : 'Available in your country');

const AlbumCard = ({
  name,
  artists,
  images,
  available_markets,
  href,
  details,
}) => (
  <div className="albumCard">
    <Image url={pathOr('http://via.placeholder.com/200x100', ['0', 'url'], images)} alt="album" />
    <div className="albumCard__info">
      <div className="albumCard__info__title">
        <AttributeField text={name} />
      </div>
      <div className="albumCard__info__extra">
        <AttributeField label="Artist" text={checkArtists(artists)} />
        <AttributeField label="Market Availability" text={checkAvailable('BR', available_markets)} />
      </div>
      <div className="albumCard__info__actions">
        <Button onClick={() => details(`${href}/tracks`)}>MORE DETAILS</Button>
      </div>
    </div>
  </div>
);

AlbumCard.propTypes = {
  name: PropTypes.string.isRequired,
  artists: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
  })).isRequired,
  images: PropTypes.arrayOf(PropTypes.shape({
    url: PropTypes.string.isRequired,
  })).isRequired,
  available_markets: PropTypes.arrayOf(PropTypes.string).isRequired,
  href: PropTypes.string.isRequired,
  details: PropTypes.func.isRequired,
};

export default AlbumCard;
