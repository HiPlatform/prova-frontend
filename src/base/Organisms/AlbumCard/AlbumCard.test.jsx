import React from 'react';
import { shallow } from 'enzyme';
import AlbumCard from './index';
import { DEFAULT_TEXTS } from '../../../consts';

describe('AlbumCard', () => {
  const details = jest.fn();
  const props = {
    name: 'My Album',
    artists: [{ name: 'Artist #1' }],
    images: [{ url: 'https://some.img/url' }],
    available_markets: ['BR'],
    href: 'https://some.redirect/url',
    details,
  };
  const component = shallow(<AlbumCard {...props} />);

  it('should render without crashing', () => {
    expect(component).toMatchSnapshot();
  });

  it('should call the details function when the "more details" button is clicked', () => {
    component.find('Button').simulate('click');
    expect(details).toHaveBeenCalled();
  });

  it('should put the first image url in card header', () => {
    expect(component.find('Image').dive().prop('src')).toBe(props.images[0].url);
  });

  it('should put a mock image url in card header if prop images has no images at all', () => {
    const images = [];
    component.setProps({ images });

    expect(component.find('Image').dive().prop('src')).toBe(DEFAULT_TEXTS.IMAGE);
  });

  it('should render the album name given in the props', () => {
    expect(component.find('.albumCard__info__title').render().text()).toBe(props.name);
  });

  it('should render the artist name', () => {
    const artistElement = component.find('.albumCard__info__extra')
      .find({ label: 'Artist' })
      .render()
      .find('p');
    expect(artistElement.text()).toBe(props.artists[0].name);
  });

  it('should render "Various artists" when the the artists prop is greater than 1', () => {
    const artists = [
      { name: 'Artist #1' },
      { name: 'Artist #2' },
    ];
    component.setProps({ artists });
    const artistElement = component.find('.albumCard__info__extra')
      .find({ label: 'Artist' })
      .render()
      .find('p');

    expect(artistElement.text()).toBe(DEFAULT_TEXTS.ARTISTS);
  });

  it('should render "Available in your country" when "BR" value is present in available_markets prop', () => {
    const marketElement = component.find('.albumCard__info__extra')
      .find({ label: 'Market Availability' })
      .render()
      .find('p');
    expect(marketElement.text()).toBe(DEFAULT_TEXTS.COUNTRY_AVAILABLE);
  });

  it('should render "Unavailable in your country" when "BR" value isn\'t present in available_markets prop', () => {
    const markets = [
      'EU', 'US', 'CA',
    ];
    component.setProps({ available_markets: markets });
    const marketElement = component.find('.albumCard__info__extra')
      .find({ label: 'Market Availability' })
      .render()
      .find('p');

    expect(marketElement.text()).toBe(DEFAULT_TEXTS.COUNTRY_UNAVAILABLE);
  });
});
