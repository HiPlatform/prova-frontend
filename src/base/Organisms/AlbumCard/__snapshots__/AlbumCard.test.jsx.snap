// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumCard should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AlbumCard
    artists={
        Array [
            Object {
              "name": "Artist #1",
            },
          ]
    }
    available_markets={
        Array [
            "BR",
          ]
    }
    details={[Function]}
    href="https://some.redirect/url"
    images={
        Array [
            Object {
              "url": "https://some.img/url",
            },
          ]
    }
    name="My Album"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Component)
          item={undefined}
          list="album"
/>,
        <Image
          alt="album"
          url="https://some.img/url"
/>,
        <div
          className="albumCard__info"
>
          <div
                    className="albumCard__info__title"
          >
                    <AttributeField
                              label={null}
                              text="My Album"
                    />
          </div>
          <div
                    className="albumCard__info__extra"
          >
                    <AttributeField
                              label="Artist"
                              text="Artist #1"
                    />
                    <AttributeField
                              label="Market Availability"
                              text="Available in your country"
                    />
          </div>
          <div
                    className="albumCard__info__actions"
          >
                    <Button
                              modifier=""
                              onClick={[Function]}
                    >
                              MORE DETAILS
                    </Button>
          </div>
</div>,
      ],
      "className": "albumCard",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "item": undefined,
          "list": "album",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "alt": "album",
          "url": "https://some.img/url",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="albumCard__info__title"
>
              <AttributeField
                            label={null}
                            text="My Album"
              />
</div>,
            <div
              className="albumCard__info__extra"
>
              <AttributeField
                            label="Artist"
                            text="Artist #1"
              />
              <AttributeField
                            label="Market Availability"
                            text="Available in your country"
              />
</div>,
            <div
              className="albumCard__info__actions"
>
              <Button
                            modifier=""
                            onClick={[Function]}
              >
                            MORE DETAILS
              </Button>
</div>,
          ],
          "className": "albumCard__info",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <AttributeField
                label={null}
                text="My Album"
/>,
              "className": "albumCard__info__title",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "label": null,
                "text": "My Album",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <AttributeField
                  label="Artist"
                  text="Artist #1"
/>,
                <AttributeField
                  label="Market Availability"
                  text="Available in your country"
/>,
              ],
              "className": "albumCard__info__extra",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "label": "Artist",
                  "text": "Artist #1",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "label": "Market Availability",
                  "text": "Available in your country",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Button
                modifier=""
                onClick={[Function]}
>
                MORE DETAILS
</Button>,
              "className": "albumCard__info__actions",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "MORE DETAILS",
                "modifier": "",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "MORE DETAILS",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Component)
            item={undefined}
            list="album"
/>,
          <Image
            alt="album"
            url="https://some.img/url"
/>,
          <div
            className="albumCard__info"
>
            <div
                        className="albumCard__info__title"
            >
                        <AttributeField
                                    label={null}
                                    text="My Album"
                        />
            </div>
            <div
                        className="albumCard__info__extra"
            >
                        <AttributeField
                                    label="Artist"
                                    text="Artist #1"
                        />
                        <AttributeField
                                    label="Market Availability"
                                    text="Available in your country"
                        />
            </div>
            <div
                        className="albumCard__info__actions"
            >
                        <Button
                                    modifier=""
                                    onClick={[Function]}
                        >
                                    MORE DETAILS
                        </Button>
            </div>
</div>,
        ],
        "className": "albumCard",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "item": undefined,
            "list": "album",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "alt": "album",
            "url": "https://some.img/url",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="albumCard__info__title"
>
                <AttributeField
                                label={null}
                                text="My Album"
                />
</div>,
              <div
                className="albumCard__info__extra"
>
                <AttributeField
                                label="Artist"
                                text="Artist #1"
                />
                <AttributeField
                                label="Market Availability"
                                text="Available in your country"
                />
</div>,
              <div
                className="albumCard__info__actions"
>
                <Button
                                modifier=""
                                onClick={[Function]}
                >
                                MORE DETAILS
                </Button>
</div>,
            ],
            "className": "albumCard__info",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <AttributeField
                  label={null}
                  text="My Album"
/>,
                "className": "albumCard__info__title",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "label": null,
                  "text": "My Album",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <AttributeField
                    label="Artist"
                    text="Artist #1"
/>,
                  <AttributeField
                    label="Market Availability"
                    text="Available in your country"
/>,
                ],
                "className": "albumCard__info__extra",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "label": "Artist",
                    "text": "Artist #1",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "label": "Market Availability",
                    "text": "Available in your country",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  modifier=""
                  onClick={[Function]}
>
                  MORE DETAILS
</Button>,
                "className": "albumCard__info__actions",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "MORE DETAILS",
                  "modifier": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "MORE DETAILS",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
